File: periodo-1/ip/projeto-1/presenter.py
import datetime
from os import system
from textwrap import dedent
from re import match

from datasource import datasource
from helpers import vinput, is_float, is_int, is_str

class Presenter:
    available_view_choices = [1,2,3,4,5,6]

    def bootstrap(self):
        self.home_view()


    def choose_view_handler(self):
        choice = int(input("(7 -> home) > "))

        if choice == 0:
            datasource.persists()
            exit(1)

        choices = {
            1: 'list_view',
            2: 'search_view',
            3: 'details_view',
            4: 'create_view',
            5: 'update_view',
            6: 'delete_view',
            7: 'home_view',
        }

        if choice not in choices.keys():
            print('Opção indisponível')
        else:
            getattr(self, choices[choice])()
        
        self.choose_view_handler()


    def home_view(self):
        system('clear')

        print(dedent("""
        ###############################
        #                             #
        #        Seja Bem-vindo       #
        #                             #
        #      Gestão de Produtos     #
        #                             #
        ###############################
        #                             #
        #     Selecione uma opção     #
        #                             #
        ###############################

        1 - Listar produtos cadastrados
        2 - Filtrar produtos cadastrados
        3 - Detalhes de um produto cadastrado
        4 - Cadastrar um novo produto
        5 - Editar um produto existente
        6 - Excluir um produto existente
        7 - Ir para a home
        """))

        self.choose_view_handler()


    def list_view(self):
        system('clear')

        print(dedent("""
        ###############################
        #                             #
        #   Lista de Produtos         #
        #                             #
        ###############################
        """))

        items = datasource.get_all()

        if len(items):
            for item in items:
                print(f"ID: {item['id']:<6} Nome: {item['name']:<40} Preço: R$ {str(item['price']).replace('.', ','):<15}")
        else:
            print("Estoque vazio\n")

        self.choose_view_handler()


    def details_view(self):
        system('clear')

        print(dedent("""
        ###############################
        #                             #
        #   Detalhes de um Produto    #
        #                             #
        #   Digite o ID do produto a  #
        #        ser visualizado      #
        ###############################

        """))

        item_id = int(input("Digite o ID do item: "))

        filtered_items = datasource.get_by_id(item_id)

        if len(filtered_items):
            item = filtered_items[0]

            print(f"\nProduto {item_id} encontrado com sucesso!")

            print(f"\nNome: {item['name']}")
            print(f"Descrição: {item['description']}")
            print(f"Preço: {float(item['price']):.2f}")
            print(f"Data de Validade: {datetime.strfdate(item['expiration_date'], '%d/%m/%Y')}")
        else:
            print("Item inexistente!")

        self.choose_view_handler()


    def search_view(self):
        system('clear')

        print(dedent("""
        ###############################
        #                             #
        #    Filtragem de produtos    #
        #                             #
        # Selecione o filtro desejado #
        #                             #
        ###############################

        1 - ID
        2 - Nome
        3 - Descrição
        4 - Preço
        5 - Data de validade

        """))

        chosen_filter = int(input("Filtrar pela propriedade: "))

        filter_choices = {
            1: "get_by_id",
            2: "get_by_name",
            3: "get_by_description",
            4: "get_by_price",
            5: "get_by_expiration_date",
        }

        if chosen_filter not in filter_choices.keys():
            print('Opção indisponível')
        else:
            filter_value = input("Digite o valor do filtro: ")
            
            if chosen_filter == 1:
                filter_value = int(filter_value)
            elif chosen_filter == 4:
                filter_value = float(filter_value)

            filtered_items = getattr(datasource, filter_choices[chosen_filter])(filter_value)

            if len(filtered_items):
                for item in filtered_items:
                    print(f"ID: {item['id']:<6} Nome: {item['name']:<40} Preço: R$ {str(item['price']).replace('.', ','):<15}")
            else:
                print("\nNão foram encontrados produtos para o filtro informado.\n")

        self.choose_view_handler()


    def create_view(self):
        system('clear')

        print(dedent("""
        ###############################
        #                             #
        #   Cadastro de Produto       #
        #                             #
        ###############################

        """))

        data = dict()

        data['name'] = vinput("Nome do Produto: ", validator=lambda v: is_str(v) and len(v) > 0)
        data['description'] = vinput("Descrição do Produto: ", validator=lambda v: is_str(v) and len(v) > 0)
        data['price'] = float(vinput("Preço do Produto: ", validator=lambda v: is_float(v) and float(v) > 0))
        data['expiration_date'] = vinput(f"Data de Validade do Produto (##/##/####): ", validator=lambda v: is_str(v) and len(v) > 0 and match(r"\d{2}/\d{2}/\d{4}", v))

        datasource.create(data)

        print("\nNovo produto cadastrado com sucesso!")

        self.choose_view_handler()


    def update_view(self):
        system('clear')

        print(dedent("""
        ###############################
        #                             #
        #   Edição de Produto         #
        #                             #
        ###############################

        """))

        data = dict()

        item_id = int(vinput("ID do Produto: "), validator=lambda v: is_int(v) and (0 < int(v) <= 100000))
        data['name'] = vinput("Nome do Produto: ", validator=lambda v: is_str(v) and len(v) > 0)
        data['description'] = vinput("Descrição do Produto: ", validator=lambda v: is_str(v) and len(v) > 0)
        data['price'] = float(vinput("Preço do Produto: "), validator=lambda v: is_float(v) and float(v) > 0)
        data['expiration_date'] = vinput(f"Data de Validade do Produto (##/##/####): ", validator=lambda v: is_str(v) and len(v) > 0 and match(v, r"\d{2}/\d{2}/\d{4}"))

        datasource.update(item_id, data)

        print("\nNovo produto atualizado com sucesso!")

        self.choose_view_handler()


    def delete_view(self):
        system('clear')

        print(dedent("""
        ###############################
        #                             #
        #   Deletar produto           #
        #                             #
        ###############################

        """))
        int(vinput("Digite o ID do produto a ser deletado: "), validator=lambda v: is_int(v) and (0 < int(v) <= 100000))

        item_id = int(vinput("Digite o ID do produto a ser deletado: "), validator=lambda v: is_int(v) and (0 < int(v) <= 100000))

        datasource.delete(item_id)

        print("\nProduto deletado com sucesso!")

        self.choose_view_handler()


presenter = Presenter()

File: periodo-2/aed1/trabalho-1/modules/presenter.c
#include <stdio.h>
#include <stdlib.h>
#include "datasource.h"
#include "presenter.h"

void escolher_tela() {
    printf("Selecione uma tela: ");

    int tela_selecionada;

    scanf("%d", &tela_selecionada);

    switch(tela_selecionada) {
        case 1: system("clear"); listar_todos(lista); break;

        case 2: system("clear"); tela_listagem_disponiveis(); break;

        case 3: system("clear"); tela_listagem_emprestados(); break;

        case 4: system("clear"); tela_listagem_por_editora(); break;

        // case 5: {

        // } break;

        // case 6: {

        // } break;

        // case 7: {

        // } break;

        case 8: system("clear"); tela_inicial(); break;

        default: printf("Opção indisponível\n");
    }

    escolher_tela();
}

void tela_inicial() {
    printf("###############################\n");
    printf("#                             #\n");
    printf("#        Seja Bem-vindo       #\n");
    printf("#                             #\n");
    printf("#      Gestão de Produtos     #\n");
    printf("#                             #\n");
    printf("###############################\n");
    printf("#                             #\n");
    printf("#     Selecione uma opção     #\n");
    printf("#                             #\n");
    printf("###############################\n\n");
    printf("1 - Listar todos os livros\n");
    printf("2 - Listar livros disponíveis\n");
    printf("3 - Listar livros emprestados\n");
    printf("4 - Listar livros por editora\n");
    printf("5 - Cadastrar um novo livro\n");
    printf("6 - Excluir um livro existente\n");
    printf("7 - Detalhes de um produto cadastrado\n");
    printf("8 - Ir para a home\n\n");
}

void tela_listagem_todos() {
    if (lista.qtd == 0) {
        printf("Não existem livros cadastrados");

        return;
    }

    listar_todos();
}

void tela_listagem_disponiveis() {
    Livro* lista_disponiveis = NULL;
    
    listar_disponiveis(&lista_disponiveis);

    if (lista_disponiveis != NULL) {
        printf("Não existem livros disponiveis\n");

        return;
    }

    printf("ID: %-4s Titulo: %-16s Editora: %-20s\n", " ", " ", " ");

    for (int i = 0; i < sizeof(lista_disponiveis) / sizeof(lista_disponiveis[0]); i++) {
        printf(
            "%-8d %-24s %-23s\n",
            lista_disponiveis[i].id,
            lista_disponiveis[i].titulo,
            lista_disponiveis[i].editora
        );
    }

    free(lista_disponiveis);
}

void tela_listagem_emprestados() {
    Livro* lista_emprestados = NULL;
    
    listar_emprestados(&lista_emprestados);

    if (lista_emprestados != NULL) {
        printf("Não existem livros emprestados\n");

        return;
    }

    printf("ID: %-4s Titulo: %-16s Editora: %-20s\n", " ", " ", " ");

    for (int i = 0; i < sizeof(lista_emprestados) / sizeof(lista_emprestados[0]); i++) {
        printf(
            "%-8d %-24s %-23s\n",
            lista_emprestados[i].id,
            lista_emprestados[i].titulo,
            lista_emprestados[i].editora
        );
    }

    free(lista_emprestados);
}

void tela_listagem_por_editora() {
    char editora[50];

    printf("Digite o nome da editora: \n");
    scanf("%s", editora);

    Livro* lista_por_editora = NULL;
    
    listar_por_editora(&lista_por_editora, editora);

    if (lista_por_editora != NULL) {
        printf("Não existem livros para a editora %s\n", editora);

        return;
    }

    printf("ID: %-4s Titulo: %-16s\n", " ", " ");

    for (int i = 0; i < sizeof(lista_por_editora) / sizeof(lista_por_editora[0]); i++) {
        printf(
            "%-8d %-24s\n",
            lista_por_editora[i].id,
            lista_por_editora[i].titulo
        );
    }

    free(lista_por_editora);
}
void tela_cadastro();
void tela_remocao();
void tela_listar_todos();
void tela_listar_disponiveis();
void tela_listar_emprestados();
void tela_listar_por_editora();
void tela_listar_por_ano();